src/Npgsql/NpgsqlConnector.cs

```diff
@@ -486,6 +489,18 @@ internal async Task Open(NpgsqlTimeout timeout, bool async, CancellationToken ca
                             Log.Error("Exception bubbled out of multiplexing read loop", t.Exception!, Id);
                         }, TaskContinuationOptions.OnlyOnFaulted);
                 }
+
+                if (_isKeepAliveEnabled)
+                {
+                    // Start the keep alive mechanism to work by scheduling the timer.
+                    // Otherwise, it doesn't work for cases when no query executed during
+                    // the connection lifetime in case of a new connector.
+                    lock (this)
```

Comment:
The only reason I see for a lock, is as a failsafe against of the connector's break from the keepalive. But by this point it's already disposed.
On the other hand, we have `GenerateResetMessage` just above, which uses the connector's `WriteBuffer`. As keepalive might already be performing, we probably should lock it.

Reply:
Yeah, I agree we should lock - there's no downside to it whatsoever (no perf impact).

We should review all locking in 6.0 though...